kernel/memory_structs/src/lib.rs

```diff
@@ -392,17 +499,12 @@ macro_rules! implement_page_frame_range {
                         
             #[doc = "A range of [`" $chunk "`]s that are contiguous in " $desc " memory."]
             #[derive(Clone, PartialEq, Eq)]
-            pub struct $TypeName(RangeInclusive<$chunk>);
+            pub struct $TypeName<P: PageSize = Page4KiB>(RangeInclusive<$chunk::<P>>);
 
             impl $TypeName {
```

Comment:
the reason these implementations are difficult is because you didn't parameterize this impl block with the `P` parameter. You need that here in order to create `PageRange`s of huge pages. Right now you only allow for creating ranges of 4K normal pages.

Once you do this, you'll see how the `empty()` function can be properly implemented (the starting page will have its `number` be `1 * P::CONTAINED_4K` instead of `11`, which doesn't make sense). Same goes for all the other functions.